tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,1176 @@
+	 */
+	cratio = rpool->ratio;
+	rcu_assign_pointer(rpool->ratio, nratio);
+	call_rcu(&cratio->rcu, tfw_sched_ratio_rpool_put);
+
+rearm:
+	smp_mb();
+	if (atomic_read(&rpool->rearm))
+		mod_timer(&rpool->timer, jiffies + interval);
+}
+
+/**
+ * Periodic function for Dynamic Ratio Scheduler.
+ */
+static void
+tfw_sched_ratio_dynamic_tmfn(unsigned long tmfn_data)
+{
+	tfw_sched_ratio_calc_tmfn((TfwSrvGroup *)tmfn_data,
+				   tfw_sched_ratio_calc_dynamic);
+}
+
+/**
+ * Periodic function for Predictive Ratio Scheduler.
+ */
+static void
+tfw_sched_ratio_predict_tmfn(unsigned long tmfn_data)
+{
+	tfw_sched_ratio_calc_tmfn((TfwSrvGroup *)tmfn_data,
+				   tfw_sched_ratio_calc_predict);
+}
+
+/*
+ * Determine if it's the turn of the server described by the server
+ * data entry at index @csidx.
+ *
+ * It's the turn of server at @csidx if sums of ratios to the left and
+ * to the right of this entry are proportional to the current iteration.
+ * As the scheduler algorithm moves forward, the sum of ratios on the
+ * left side decreases. When a server is selected, its current ratio
+ * is decremented, so the sum of ratios decreases by 1 as well.
+ *
+ * With that in mind, ratios that have a huge difference should not be
+ * specified for servers in the same group. A decrement of a huge sum
+ * would be too insignificant to affect the scheduling algorithm. Thus
+ * weights like { 10, 1 } make more sense than weights like { 1000, 10 }.
+ * Requests are distributed proportionally in both cases, but significant
+ * bursts are possible in the first case.
+ *
+ * TODO: The algorithm may and should be improved.
+ */
+static inline bool
+tfw_sched_ratio_is_srv_turn(TfwRatioPool *rpool, TfwRatio *ratio, size_t csidx)
+{
+	unsigned long headsum2, tailsum2;
+	TfwRatioSrvData *srvdata = ratio->srvdata;
+	TfwRatioSchData *schdata = &ratio->schdata;
+
+	if (!csidx)
+		return true;
+
+	headsum2 = (srvdata[0].cratio + srvdata[csidx - 1].cratio) * csidx;
+	tailsum2 = (srvdata[csidx].cratio
+		    + (srvdata[rpool->srv_n - 1].cratio
+		       ? : srvdata[rpool->srv_n - 1].oratio))
+		   * (rpool->srv_n - csidx);
+
+	return tailsum2 * schdata->riter > headsum2;
+}
+
+/*
+ * Get the index of the next server descriptor.
+ *
+ * The array of server data entries used by the algorithm must be sorted
+ * by ratio in descending order, with the higher weight entries moved
+ * towards the start of the array.
+ *
+ * For concurrent use the algorithm is synchronized by a plain spin lock.
+ * A lock-free implementation of the algorithm as it is would require too
+ * many atomic operations including CMPXCHG and checking loops. It seems
+ * that it won't give any advantage.
+ */
+static TfwRatioSrvDesc *
+tfw_sched_ratio_next_srv(TfwRatioPool *rpool, TfwRatio *ratio)
+{
+	size_t csidx;
+	TfwRatioSrvData *srvdata = ratio->srvdata;
+	TfwRatioSchData *schdata = &ratio->schdata;
+
+	/* Start with server that has the highest ratio. */
+	spin_lock(&schdata->lock);
```

Comment:
This is a bad thing. We have to use the lock as described in the comment, but we still want to have RR (ratio) scheduler as the most quick one. However, probably that would be an issue now if we for example try it to run on 4 processors machine (e.g. 40 and more CPU cores). Thus, let's introduce 2 versions of the function: this one for dynamic ratios and the old one, from RR scheduler, for static ratios.